docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,387 @@
+  arguments ``weights_init`` and ``biases_init`` of the constructor) to the
+  children bricks.
+* :meth:`.Brick.get_dim`: implementing this function is useful if you want
+  to provide a simple way to get the dimensions of the inputs and outputs of
+  the brick.
+
+If you want to inherit from a specific brick, check its docstring to
+identify the particular methods to overwrite and the attributes to define.
+
+Application methods
+~~~~~~~~~~~~~~~~~~~
+
+The :meth:`apply` method listed above is probably the most
+important method of your brick because it is the one that actually takes
+theano tensors as inputs, process them and return output tensors. You should
+decorate it with the :func:`.application` decorator, which names variables
+and register auxiliary variables of the operation you implement.
+It is used as follows:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application(inputs=['input1', 'input2'], outputs=['output'])
+    ...     def apply(self, input1, input2)
+    ...         ...
+    ...         return something
+
+In the case above, it will automatically label the theano tensor variable
+``input1`` to ``Foo_apply_input1``, idem for ``input2`` and the output of the
+method.
+
+Under the hood, the ``@application`` decorator creates an object of class
+:class:`.Application`, named ``apply``, which becomes an attribute of the
+brick class (by opposition to class instances).
+
+Application properties
+""""""""""""""""""""""
+
+In the previous examples, the names of the arguments of the application methods
+were directly provided as arguments of the ``@application`` decorator because
+they were common to all instances of the classes. On the other hand, if these
+names need to be defined differently for particular instances of the class,
+you should use the ``apply.property`` decorator. Let's say that we want to
+name our attribute inputs with the string ``self.fancy_name``, then we should
+write:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     def __init__(self, fancy_name):
+    ...         self.fancy_name = fancy_name
+    ...     @application
+    ...     def apply(self, input)
+    ...         ...
+    ...     @apply.property('inputs')
+    ...     def apply_inputs(self):
+    ...         # Note that you can use any python code to define the name
+    ...         return self.fancy_name
+
+Using application calls
+"""""""""""""""""""""""
+
+If you need to add auxiliary variables to the computation graph
+:class:`.ComputationGraph`, you need to pass ``application_call`` as argument
+of your ``apply`` function, as shown in this example:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application
+    ...     def apply(self, x, application_call):
+    ...         application_call.add_auxiliary_variable(x.mean())
+    ...         return x + 1
+
+You can retrieve these annotated variables as usual with the computational
+graph :class:`.ComputationGraph` and filters :class:`.VariableFilter`.
```

Comment:
Please add an example and please say one sentence why one might need auxiliary variables at all.